app/logic/minor.py

```diff
@@ -68,34 +111,171 @@ def getProgramEngagementHistory(program_id, username, term_id):
+                                           .where(IndividualRequirement.username.is_null(True),
+                                                  CertificationRequirement.certification == Certification.CCE, 
+                                                  CertificationRequirement.name.not_in(['Summer Program'])))
+    if action == 'add':
+        try: 
+            IndividualRequirement.create(**{engagementData['type']: engagementData['id'],
+                                           "username": engagementData['username'],
+                                           "term": engagementData['term'],
+                                           "requirement": requirement.get(),
+                                           "addedBy": currentUser,
+                                        })
+            
+        except DoesNotExist as e:
+            raise e 
+        
+    elif action == 'remove':
+        IndividualRequirement.delete().where(
+                getattr(IndividualRequirement, engagementData['type']) == engagementData['id'],
+                IndividualRequirement.username == engagementData['username'],
+                IndividualRequirement.term == engagementData['term']
+            ).execute()
+    else:
+        raise Exception(f"Invalid action '{action}' sent to setCommunityEngagementForUser")
 
 
 def getCommunityEngagementByTerm(username):
     """
     Given a username, return all of their community engagements (service learning courses and event participations.)
     """
-    courses = (Course.select(Course)
+    courseMatchCase =  Case(None, [(IndividualRequirement.course.is_null(True) , 0)], 1)
+
+    courses = (Course.select(Course, courseMatchCase.alias("reqMatch"))
                      .join(CourseParticipant, on=(Course.id == CourseParticipant.course))
+                     .join(IndividualRequirement, JOIN.LEFT_OUTER, on=((IndividualRequirement.course == Course.id) & 
+                                                                       (IndividualRequirement.username == CourseParticipant.user) & 
+                                                                       (IndividualRequirement.term == Course.term)))
                      .where(CourseParticipant.user == username)
                      .group_by(Course.courseName, Course.term))
     
     # initialize default dict to store term descriptions as keys mapping to each
     # engagement's respective type, name, id, and term.
     terms = defaultdict(list)
     for course in courses:
-        terms[(course.term.description, course.term.id)].append({"name":course.courseName, "id":course.id, "type":"course", "term":course.term})
+        terms[(course.term.description, course.term.id)].append({
+                    "name":course.courseName, 
+                    "id":course.id, 
+                    "type":"course", 
+                    "matched": course.reqMatch,
+                    "term":course.term.id})
```

Comment:
I don't like the formatting here, Anderson doesn't either. Let's either add an indentation before the closing of the dict or remove the indentation after the append.